### PR TITLE
Fix #176: Training, validation an test each in separate sessions and own graphs 

### LINKED ISSUE #176: Training, validation an test each in separate sessions and own graphs

### PR DESCRIPTION
**Pull request by tilmankamp**

### CODE REVIEW

**Comment by Cwiiis** [comment]
First comment pre-review, I think this should be squashed if the commits don't work alone.

**Comment by tilmankamp** [comment]
@Cwiiis That's right - here you go.

File: DeepSpeech.ipynb [hunk 1]
```diff
@@ -1320,170 +1292,390 @@
...
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "def stop_execution_context(execution_context, session, checkpoint_path=None, global_step=None):\n",
+    "    \n",
+    "    # If the model is not persisted, we'll return 'None'\n",
+    "    hibernation_path = None\n",
+    "    \n",
+    "    if checkpoint_path is not None and global_step is not None:\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: I think this would read easier as `if checkpoint_path and global_step:`

File: DeepSpeech.ipynb [hunk 2]
```diff
@@ -1320,170 +1292,390 @@
...
+    "    \n",
+    "    return hibernation_path"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now let's introduce the main routine for training and inference. \n",
+    "It takes a started execution context (given by `execution_context` and `session`), an optional `epoch` index and a flag wheather to calculate WER report data (`query_report`) or not.\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: s/wheather/whether/

File: DeepSpeech.ipynb [hunk 3]
```diff
@@ -1320,170 +1292,390 @@
...
+    "Its main duty is to iterate over all batches and calculate the mean loss.\n",
+    "If an epoch is provided, it will also do back propagation - aka training.\n",
+    "It will return a batch set result tuple which consists of the mean loss and - if asked for report data - the mean accuracy and individual results tuple."
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: I'm having trouble scanning this sentence, definitely needs some commas or possible rephrasing... Rather than having a variable return value, could the `query_report` part just be a separate function?

The function is a bit hard to follow because of all the conditional paths right now, it'd be good if we could decouple it somewhat, or at least rearrange it so that the paths don't have to be interleaved quite so much.

File: DeepSpeech.ipynb [hunk 4]
```diff
@@ -1320,170 +1292,390 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "def apply_batches(execution_context, session, epoch=-1, query_report=False):\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: Why is this called `apply_batches`? I think a name that's more descriptive of its function would be good. Something like `calculate_loss` (and splitting out the reporting would make this a more accurate name), or assuming the split isn't feasible, `calculate_loss_and_wer_report` or something?

File: DeepSpeech.ipynb [hunk 5]
```diff
@@ -1320,170 +1292,390 @@
...
+    "    batches_per_device = ceil(float(data_set.total_batches) / len(available_devices))\n",
     "\n",
-    "    # Apply gradients to modify the model\n",
-    "    apply_gradient_op = apply_gradients(optimizer, avg_tower_gradients)\n",
+    "    total_loss = 0.0\n",
+    "    params = [avg_loss]\n",
+    "    \n",
+    "    # Facilitate a train run\n",
+    "    if do_training:\n",
+    "        params.append([apply_gradient_op, merged])\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: If this was shuffled a little so that tower_results was called before unpacking variables and params set after that, we wouldn't need this extra if statement - it'd be good if we could consolidate some of these many if's, for readability.

File: DeepSpeech.ipynb [hunk 6]
```diff
@@ -1320,170 +1292,390 @@
...
+    "\n",
+    "        # Compute the batch\n",
+    "        result = session.run(params, **extra_params)\n",
+    "        \n",
+    "        # Offset to read from result list\n",
+    "        offset = 0\n",
+    "        \n",
+    "        # Add batch to loss\n",
+    "        total_loss += result[offset]\n",
+    "        offset = offset + 1\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: may as well just `offset += 1` for brevity?

File: DeepSpeech.ipynb [hunk 7]
```diff
@@ -1320,170 +1292,390 @@
...
     "\n",
-    "    # Create a saver to checkpoint the model\n",
-    "    saver = tf.train.Saver(tf.all_variables())\n",
+    "    # Loop over the batches\n",
+    "    for batch in range(int(batches_per_device)):\n",
+    "        extra_params = { }\n",
+    "        if do_training and do_fulltrace:\n",
+    "            loss_run_metadata            = tf.RunMetadata()\n",
+    "            extra_params['options']      = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)\n",
+    "            extra_params['run_metadata'] = loss_run_metadata\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: Rather than doing this whole block for every iteration, couldn't `extra_params` and `extra_params['options']` be initialised outside of the loop, and take out a bit of work here?

File: DeepSpeech.ipynb [hunk 8]
```diff
@@ -1320,170 +1292,390 @@
...
+    "        offset = offset + 1\n",
+    "        \n",
+    "        if do_training:\n",
+    "            # Log all variable states in current step\n",
+    "            step = epoch * data_set.total_batches + batch * len(available_devices)\n",
+    "            writer.add_summary(result[offset][1], step)\n",
+    "            if do_fulltrace:\n",
+    "                writer.add_run_metadata(loss_run_metadata, 'loss_epoch%d_batch%d'   % (epoch, batch))\n",
+    "            writer.flush()\n",
+    "            offset = offset + 1\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: Same here with the incrementing.

File: DeepSpeech.ipynb [hunk 9]
```diff
@@ -1320,170 +1292,390 @@
...
+    "        return (loss, total_accuracy / batches_per_device, report_results)\n",
+    "    else:\n",
+    "        return (loss, None, None)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The following routine will print a report from a provided batch set result tuple. It takes a caption for titling the output plus the batch set result tuple. If the batch set result tuple contains accuracy and a report results tuple, a complete WER report will be calculated, printed and its mean WER returned. Otherwise it will just print the loss and return `None`."
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: This function doesn't look like it needs to do two things conditionally - the second thing could just be a separate function.

File: DeepSpeech.ipynb [hunk 10]
```diff
@@ -1320,170 +1292,390 @@
...
+    "        print title\n",
+    "\n",
+    "    return mean_wer"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's also introduce a routine that facilitates a complete run of a batch set (given by its name in `set_name`) - from execution contet creation to closing the session. \n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: s/contet/content/

Comment (Cwiiis) [inline DeepSpeech.ipynb]: I think a 'complete run of a batch set' doesn't make sense. I guess, this really creates execution contexts and constructs graphs to be used when doing inference in aid of a training/validation/test run, but the names of these functions and some of the interleaving makes this quite unclear I think. This needs some looking at I think.

File: DeepSpeech.ipynb [hunk 11]
```diff
@@ -1320,170 +1292,390 @@
...
     "    \n",
     "    # Loop over the data set for training_epochs epochs\n",
     "    for epoch in range(start_epoch, training_iters):\n",
+    "\n",
+    "        print \"STARTING Epoch\", '%04d' % (epoch)\n",
     "        \n",
-    "        print \"\\n\\n\\nSTARTING Epoch\", '%04d' % (epoch)\n",
+    "        if epoch == 0 or hibernation_path is not None:\n",
+    "            if hibernation_path is not None:\n",
+    "                print \"Resuming training session from\", \"%s\" % hibernation_path, \"...\"\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: Maybe it'd make more sense for this print statement to be in `start_execution_context` (I guess it couldn't be quite the same statement then, but there's an amount of redundancy here that makes things harder to read than they need to be).

File: DeepSpeech.ipynb [hunk 12]
```diff
@@ -1320,170 +1292,390 @@
...
-    "                \"Training avg_cer=\" + \"{:.9f}\".format((total_accuracy / ceil(batches_per_device))), train_results)\n",
-    "            print\n",
+    "        print \"Training model...\"\n",
+    "        train_time = stopwatch(train_time)\n",
+    "        result = apply_batches(train_context, session, epoch=epoch, query_report=is_display_step)\n",
+    "        train_time = stopwatch(train_time)\n",
+    "\n",
+    "        result = print_report(\"Training\", result)\n",
+    "        # If there was a WER calculated, we keep it\n",
+    "        train_wer = result if result is not None else train_wer\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: I think this is needlessly on one line, it'd read better and do less work to do
```
if result is not None:
    train_wer = result
```

File: DeepSpeech.ipynb [hunk 13]
```diff
@@ -1320,170 +1292,390 @@
...
+    "            print \"Hibernating training session into directory\", \"%s\" % checkpoint_dir\n",
     "            checkpoint_path = os.path.join(checkpoint_dir, 'model.ckpt')\n",
-    "            print \"Checkpointing in directory\", \"%s\" % checkpoint_dir\n",
-    "            saver.save(session, checkpoint_path, global_step=epoch)\n",
-    "            print\n",
+    "            # If the hibernation_path is set, the next epoch loop will load the model\n",
+    "            hibernation_path = stop_execution_context( \\\n",
+    "                train_context, session, checkpoint_path=checkpoint_path, global_step=epoch)\n",
     "            \n",
+    "            # Validating the moel in a fresh session\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: s/moel/model/

File: DeepSpeech.ipynb [hunk 14]
```diff
@@ -1320,170 +1292,390 @@
...
+    "            print \"Validating model...\"\n",
+    "            result = run_set('dev', model_path=hibernation_path, query_report=(epoch == training_iters - 1))\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: So we only want a WER report on validation on the last epoch? (should this be configurable?)

File: DeepSpeech.ipynb [hunk 15]
```diff
@@ -1320,170 +1292,390 @@
...
+    "            result = print_report(\"Validation\", result)\n",
+    "            # If there was a WER calculated, we keep it\n",
+    "            dev_wer = result if result is not None else dev_wer\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: Same here with the one-liners, this is hard to read and more work than just not doing it.

File: DeepSpeech.ipynb [hunk 16]
```diff
@@ -1320,170 +1292,390 @@
...
     "        print \"FINISHED Epoch\", '%04d' % (epoch), \\\n",
-    "            \"   Overall time:\", format_duration(overall_time), \\\n",
-    "            \"  Training time:\", format_duration(training_time), \\\n",
-    "            \"   Average loss:\", \"{:.9f}\".format((loss / ceil(batches_per_device)))\n",
+    "            \"  Overall epoch time:\", format_duration(overall_time), \\\n",
+    "            \"  Training time:\", format_duration(train_time)\n",
     "        print\n",
-    "        \n",
+    "\n",
+    "    # If the last iteration step was no validation, we still have to save the model\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: I don't understand this comment.

File: DeepSpeech.ipynb (hunk 14 above)
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: Having this only on the last epoch seems counter to the purpose of printing out a validation WER. It should be configurable or maybe happen all the time but have `is_validation_step` set to a computationally reasonable value.

Traditionally one prints out a validation WER to monitor the generalization of the model as it trains. To do this one needs to see the training WER and the validation WER throughout training to see if the training WER dips significantly below the validation WER.

When this happens you want to stop training, as this is a signal of over fitting and subsequent epochs will only server to further overfit and thus waste GPU time.

File: DeepSpeech.ipynb (hunk 16 above)
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: When validation occurs the model is serialized and `hibernation_path` is not `None`, see this line
```python
            hibernation_path = stop_execution_context( \
                train_context, session, checkpoint_path=checkpoint_path, global_step=epoch)
```
However,  if `hibernation_path` is `None`, then the model was not serialized as a result of validation and we need to do it here.

File: DeepSpeech.ipynb (hunk 12 above)
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: As `is_display_step` need not be true for the final epoch, `train_wer` will not reflect the final WER.

File: DeepSpeech.ipynb [hunk 17]
```diff
@@ -1320,170 +1292,390 @@
...
+    "            # Loading the model into the session\n",
+    "            execution_context[3].restore(session, model_path)\n",
+    "\n",
+    "        # Start queue runner threads\n",
+    "        runner_threads = tf.train.start_queue_runners(sess=session)\n",
     "\n",
-    "    # Average tower gradients\n",
-    "    avg_tower_gradients = average_gradients(tower_gradients)\n",
+    "        # Start importer's queue threads\n",
+    "        queue_threads = execution_context[1].start_queue_threads(session)\n",
```
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: Why return `queue_threads` they are never used?

File: util/importers/ldc93s1.py [hunk 18]
```diff
@@ -57,13 +57,14 @@ def start_queue_threads(self, session):
         for batch_thread in batch_threads:
             batch_thread.daemon = True
             batch_thread.start()
+        return batch_threads
```
Comment (kdavis-mozilla) [inline util/importers/ldc93s1.py]: Why return `batch_threads` they are never used.

File: util/importers/librivox.py [hunk 19]
```diff
@@ -59,17 +59,18 @@ def __init__(self, txt_files, thread_count, batch_size, numcep, numcontext):
...
         return max(len(available_gpus), 1)
-    
+
     def start_queue_threads(self, session):
         batch_threads = [Thread(target=self._populate_batch_queue, args=(session,)) for i in xrange(self._thread_count)]
         for batch_thread in batch_threads:
             batch_thread.daemon = True
             batch_thread.start()
-    
+        return batch_threads
```
Comment (kdavis-mozilla) [inline util/importers/librivox.py]: Why return `batch_threads` they are never used.

File: util/importers/ted.py [hunk 20]
```diff
@@ -75,6 +75,7 @@ def start_queue_threads(self, session):
         for batch_thread in batch_threads:
             batch_thread.daemon = True
             batch_thread.start()
+        return batch_threads
```
Comment (kdavis-mozilla) [inline util/importers/ted.py]: Why return `batch_threads` they are never used.

File: util/importers/ldc93s1.py [hunk 21]
```diff
@@ -74,11 +75,14 @@ def _compute_source_target(self):
...
-                self._x_length: source_len,
-                self._y: target,
-                self._y_length: target_len})
+            try:
+                session.run(self._enqueue_op, feed_dict={
+                    self._x: source,
+                    self._x_length: source_len,
+                    self._y: target,
+                    self._y_length: target_len})
+            except (RuntimeError, tf.errors.CancelledError):
```
Comment (kdavis-mozilla) [inline util/importers/ldc93s1.py]: Not pretty, but I don't have an easier solution.

File: util/importers/librivox.py [hunk 22]
```diff
@@ -90,11 +91,14 @@ def _populate_batch_queue(self, session):
...
-                self._x_length: source_len,
-                self._y: target,
-                self._y_length: target_len})
+            try:
+                session.run(self._enqueue_op, feed_dict={
+                    self._x: source,
+                    self._x_length: source_len,
+                    self._y: target,
+                    self._y_length: target_len})
+            except (RuntimeError, tf.errors.CancelledError):
```
Comment (kdavis-mozilla) [inline util/importers/librivox.py]: Not pretty, but I don't have an easier solution.

File: util/importers/ted.py [hunk 23]
```diff
@@ -98,11 +99,14 @@ def _populate_batch_queue(self, session):
...
-                self._x_length: source_len,
-                self._y: target,
-                self._y_length: target_len})
+            try:
+                session.run(self._enqueue_op, feed_dict={
+                    self._x: source,
+                    self._x_length: source_len,
+                    self._y: target,
+                    self._y_length: target_len})
+            except (RuntimeError, tf.errors.CancelledError):
```
Comment (kdavis-mozilla) [inline util/importers/ted.py]: Not pretty, but I don't have an easier solution.

File: DeepSpeech.ipynb (hunk 5 above)
Comment (Cwiiis) [inline DeepSpeech.ipynb]: I mean if you put line 1450 and line 1455 above line 1446, you could merge line 1458 with line 1446.

File: DeepSpeech.ipynb (hunk 10 above)
Comment (Cwiiis) [inline DeepSpeech.ipynb]: But you still can't 'run' a data-set, right? You perform things with/on it - I think we should be more explicit here so it can be understood in isolation.

File: DeepSpeech.ipynb (hunk 16 above)
Comment (Cwiiis) [inline DeepSpeech.ipynb]: So this would be much clearer if it read:
> If `hibernation_path` is `None`, then the model was not serialized and we need to do that here.
You could elaborate by then saying:
> This happens if validation was not performed on the preceding iteration.

File: DeepSpeech.ipynb (hunk 5 above)
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: @Cwiiis But `tower_results` and is not defined before line 1446. It is defined on line 1447 or line 1449

File: DeepSpeech.ipynb (hunk 9 above)
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: @Cwiiis I'm going to ignore this for now as it's Friday night and I want to get this merged and the run going.

File: DeepSpeech.ipynb (hunk 10 above)
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: Actually s/contet/context/

File: DeepSpeech.ipynb (hunk 5 above)
Comment (Cwiiis) [inline DeepSpeech.ipynb]: Good point, missed that. I still think some restructuring here would help readability, but it might also require some shuffling elsewhere. It's too tangled for my liking - perhaps it would help if we defined some classes so that we can start using readable names rather than [0], [1], or having to extract things into separate variables? (i.e. we could say tower_results.params, or even execution_context.tower_results.params, or something along these lines - alternatively, getter functions)

File: DeepSpeech.ipynb (hunk 11 above)
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: But some executions of `start_execution_context()` print out a message. Some don't.

Comment (Cwiiis) [inline DeepSpeech.ipynb]: True, on further thought, this comment is just invalid.

File: util/importers/ldc93s1.py (hunk 18 above)
Comment (kdavis-mozilla) [inline util/importers/ldc93s1.py]: Captured in issue #231

File: util/importers/librivox.py (hunk 19 above)
Comment (kdavis-mozilla) [inline util/importers/librivox.py]: Captured in issue #231

File: util/importers/ted.py (hunk 20 above)
Comment (kdavis-mozilla) [inline util/importers/ted.py]: Captured in issue #231

File: DeepSpeech.ipynb (hunk 5 above)
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: I agree with your point, but for the time being I'm just going to leave it as is. (I want to go home at some point.)

**Comment by kdavis-mozilla** [comment]
LGTM
